miniapps/autodiff/admfem.hpp

```diff
@@ -0,0 +1,587 @@
+#include <codi.hpp>
+
+namespace mfem {
+
+namespace ad {
+#ifdef MFEM_USE_ADFORWARD
+typedef codi::RealForward ADFloatType;
+typedef TADVector<ADFloatType> ADVectorType;
+typedef TADDenseMatrix<ADFloatType> ADMatrixType;
+#else
+typedef codi::RealReverse ADFloatType;
+typedef TADVector<ADFloatType> ADVectorType;
+typedef TADDenseMatrix<ADFloatType> ADMatrixType;
+#endif
+}
+
+
+template<int vector_size=1, int state_size=1, int param_size=0>
+class VectorFuncAutoDiff
+{
+public:
+    VectorFuncAutoDiff(std::function<void(mfem::Vector&, ad::ADVectorType&, ad::ADVectorType&)> F_)
+    {
+            F=F_;
+    }
+
+    void QJacobian(mfem::Vector &vparam, mfem::Vector &uu, mfem::DenseMatrix &jac)
+    {
+#ifdef MFEM_USE_ADFORWARD
+        // use forward mode
+        jac.SetSize(vector_size, state_size);
+        jac = 0.0;
+        {
+            ad::ADVectorType aduu(state_size);
+            ad::ADVectorType rr(vector_size);
+            for(int i=0;i<state_size;i++){
+                aduu[i].setValue(uu[i]);
+                aduu[i].setGradient(0.0);
+            }
+            for(int ii=0;ii<state_size;ii++){
+                aduu[ii].setGradient(1.0);
+                F(vparam,aduu,rr);
+                for(int jj=0;jj<vector_size;jj++)
+                {
+                    jac(jj,ii)=rr[jj].getGradient();
+                }
+                aduu[ii].setGradient(0.0);
+            }
+        }
+#else // use reverse mode
+        jac.SetSize(vector_size, state_size);
+        jac = 0.0;
+        {
+            ad::ADVectorType aduu(state_size);
+            ad::ADVectorType rr(vector_size);
+            for(int i=0;i<state_size;i++){
+                aduu[i]=uu[i];
+            }
+
+            ad::ADFloatType::TapeType& tape =ad::ADFloatType::getGlobalTape();
+            typename ad::ADFloatType::TapeType::Position pos=tape.getPosition();
+
+            tape.setActive();
+            for(int ii=0;ii<state_size;ii++){ tape.registerInput(aduu[ii]); }
+            F(vparam,aduu,rr);
+            for(int ii=0;ii<vector_size;ii++){ tape.registerOutput(rr[ii]); }
+            tape.setPassive();
+
+            for(int jj=0;jj<vector_size;jj++){
+                rr[jj].setGradient(1.0);
+                tape.evaluate();
+                for(int ii=0;ii<state_size;ii++){
+                    jac(jj,ii)=aduu[ii].getGradient();
+                }
+                tape.clearAdjoints();
+                rr[jj].setGradient(0.0);
+            }
+            tape.reset(pos);
+        }
+#endif
+    }
+private:
+    std::function<void(mfem::Vector&, ad::ADVectorType&, ad::ADVectorType&)> F;
+}; //VectorFuncAutoDiff
+
+
+template<template<typename, typename, typename, int, int, int> class CTD
+         , int vector_size=1, int state_size=1, int param_size=0>
+class QVectorFuncAutoDiff
+{
+public:
+    /// Evaluates the vector function for given set of parameters and state
+    /// values in vector uu. The result is returned in vector rr.
```

Comment:
Do these comments mean that `vparam` and `uu` are inputs and `rr` is the only output? If so, why isn't it `const mfem::Vector &uu`?

Reply:
I am using the same template implementation for evaluating the function and for evaluating the gradients. In the implementation uu and rr have the same type. In this case uu cannot be constant as for gradient evaluation the dual part is modified. One can split the types but such a solution will add more template parameters.

Reply:
Thanks for the explanation!